---
title: gwaspr R Package
output: github_document
always_allow_html: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = T, message = F, warning = F)
```

`gwaspr`: an `R` package for plotting GWAS results from the `GAPIT` package

# Installation

```{r eval = F}
devtools::install_github("derekmichaelwright/gwaspr")
```

```{r eval = T}
library(gwaspr)
```

![](man/figures/logo_gwaspr.png)

# GWAS Tutorial

https://derekmichaelwright.github.io/dblogr/academic/gwas_tutorial

# Usage

For best practice, output from GAPIT should be in its own folder. In this case, they are located in a folder called `GWAS_Results/`. For this example we will plot GWAS results from 3 traits in a lentil diversity panel:

- ****Cotyledon_Color**: a *qualitative* trait describing cotyledon color (Red = 0, Yellow = 1).
- **DTF_Nepal_2017**: a *quantitative* trait describing days from sowing to flowering in a 2017 Nepal field trial. 
- **DTF_Sask_2017**: a *quantitative* trait describing days from sowing to flowering in a 2017 Saskatchewan field trial. 
- **DTF_Sask_2017_b**: same as above but run with the *b* coefficient from a photothermal model (see [Wright *et al*. 2020](https://doi.org/10.1002/ppp3.10158)) used as a covariate. 

Note: for more info check out this [GWAS tutorial](https://derekmichaelwright.github.io/dblogr/academic/gwaspr_tutorial).

# gwaspr Functions

## list_Traits()

```{r eval = T}
list_Traits(folder = "GWAS_Results/")
```

## list_Results_Files()

```{r eval = T}
list_Result_Files(folder = "GWAS_Results/")
```

## is_ran()

```{r}
is_ran(folder = "GWAS_Results/")
```

## Order_GWAS_Results()

```{r eval = F}
# first reorder the result files if they are not already arranged by P.value
order_GWAS_Results(folder = "GWAS_Results/")
```

## is_ordered()

```{r}
is_ordered(folder = "GWAS_Results/")
```

## table_GWAS_Results()

```{r eval = T}
myResults <- table_GWAS_Results(folder = "GWAS_Results/",
                                threshold = 6.8, sug.threshold = 5)
myResults[1:10,]
```

## table_GWAS_Results_Summary()

```{r}
mySummary <- table_GWAS_Results_Summary(folder = "GWAS_Results/")
mySummary
```

## list_Top_Markers()

```{r eval = T}
list_Top_Markers(folder = "GWAS_Results/", trait = "DTF_Nepal_2017")
```

```{r eval = T}
list_Top_Markers(folder = "GWAS_Results/", trait = "DTF_Sask_2017")
```

```{r eval = T}
list_Top_Markers(folder = "GWAS_Results/", trait = "DTF_Sask_2017_b")
```

```{r eval = T}
list_Top_Markers(folder = "GWAS_Results/", trait = "Cotyledon_Color")
```

---

## gg_Manhattan()

### Multi-Model Manhattan Plots

This is the default setting for this function: `facet = F`

```{r eval = F}
for(i in list_Traits(folder = "GWAS_Results/")) {
  # Plot
  mp <- gg_Manhattan(folder = "GWAS_Results/", trait = i)
  # Save
  ggsave(paste0("man/figures/gg_Man1_", i, ".png"), 
         mp, width = 12, height = 3.5, bg = "white")
}
```

![](man/figures/Multi_DTF_Nepal_2017.png)

![](man/figures/Multi_DTF_Sask_2017.png)

![](man/figures/Multi_DTF_Sask_2017_CV_b.png)

![](man/figures/Multi_Cotyledon_Color.png)

---

### Facetted Manhattan Plots

facet the plots by setting `facet = T`

```{r eval = F}
for(i in list_Traits(folder = "GWAS_Results/")) {
  # Plot
  mp <- gg_Manhattan(folder = "GWAS_Results/", trait = i, facet = T)
  # Save
  ggsave(paste0("man/figures/gg_Man2_", i, ".png"), 
         mp, width = 12, height = 8, bg = "white")
}
```

![](man/figures/Facet_DTF_Nepal_2017.png)

![](man/figures/Facet_DTF_Sask_2017.png)

![](man/figures/Facet_DTF_Sask_2017_CV_b.png)

![](man/figures/Facet_Cotyledon_Color.png)

---

## gg_Mahattan_Zoom()

```{r eval = F}
# Plot
mp <- gg_Manhattan_Zoom(
  # Specify a folder with GWAS results
  folder = "GWAS_Results/", 
  # Select a trait to plot
  trait = "Cotyledon_Color", 
  # Plot just Chromosome 1
  chr = 1,
  pos1 = 360000000,
  pos2 = 370000000,
  # Highlight specific markers
  markers = c("Lcu.1GRN.Chr1p365986872",
              "Lcu.1GRN.Chr1p361840399"),
  # Create alt labels for the markers
  labels = c("365Mbp","361Mbp"),
  # Specify Color for each marker vline
  vline.colors = rep("green",3),
  # Should models be facetted
  facet = T
  )
# Save
ggsave("man/figures/gg_Man_Zoom1.png", mp, width = 8, height = 8)
```

![](man/figures/gg_Man_Zoom1.png)

```{r eval = F}
# Plot
mp <- gg_Manhattan_Zoom(
  # Specify a folder with GWAS results
  folder = "GWAS_Results/", 
  # Select a trait to plot
  trait = "Cotyledon_Color", 
  # Plot just Chromosome 1
  chr = 1,
  pos1 = 360000000,
  pos2 = 370000000,
  # Plot only certain GWAS models
  models = c("MLMM","FarmCPU","BLINK"),
  # Highlight specific markers
  markers = c("Lcu.1GRN.Chr1p365986872",
              "Lcu.1GRN.Chr1p361840399"),
  # Create alt labels for the markers
  labels = c("365Mbp","361Mbp"),
  # Specify Color for each marker vline
  vline.colors = rep("green",3),
  # Should models be facetted
  facet = F,
  # set a max P value
  pmax = 50
  )
# Save
ggsave("man/figures/gg_Man_Zoom2.png", mp, width = 8, height = 4)
```

![](man/figures/gg_Man_Zoom2.png)

---

## gg_Manhattan_Traits()

```{r eval = F}
# Plot
mp <- gg_Manhattan_Traits(
  # Specify a folder with GWAS results
  folder = "GWAS_Results/",
  # Select traits to plot
  traits = c("DTF_Sask_2017", "DTF_Nepal_2017"),
  # Specify a title
  title = "Days to Flower",
  # Highlight specific markers
  markers = c("Lcu.1GRN.Chr2p44545877",
              "Lcu.1GRN.Chr5p1658484",
              "Lcu.1GRN.Chr6p3269280"),
  # Create alt labels for the markers
  labels = c("44Mbp","16Mbp","32Mbp"),
  # Specify Color for each marker vline
  vline.colors = c("red","red","blue")
)
# Save
ggsave("man/figures/gg_Man_Traits.png", mp, width = 12, height = 8, bg = "white")
```

![](man/figures/gg_Man_Traits.png)


## gg_GWAS_Summary()

```{r eval = F}
# Plot
mp <- gg_GWAS_Summary(folder = "GWAS_Results/", 
                      title = "Summary of Significant GWAS Results")
# Save
ggsave("man/figures/GWAS_Summary_01.png", mp, width = 12, height = 4)
```

![](man/figures/GWAS_Summary_01.png)

Make it an interactive plot with the following code

```{r eval = F, echo = F}
gg_plotly(mp, filename = "man/figures/GWAS_Summary.html")
plotly::ggplotly(mp)
```

---

# GAPIT

`GAPIT`: and `R` package for performing Genome Wide Association Studies (GWAS)

https://github.com/jiabowang/GAPIT

# Dependancies

`tidyverse`, `ggpubr`, `ggbeeswarm`, `ggrepel`, `ggtext`, `plotly`, `htmlwidgets`

---

&copy; Derek Michael Wright
